### ISSUE TITLE
Augmented Image memory footprint increase (Bug?)

### ISSUE THREAD

**Issue by marvision-ai** [aleju/imgaug#709]
When applying any `augmentation` to a base image and then using opencv `cv2.imwrite()` to save the image, this results in an image **much larger**  in memory footprint. 

For example: 
image = 3840x2160 pix ==> 
**total size = 1.2mb**

augmented_image = 3840x2160 pix  
**total size = 7.5mb** (this image is augmented randomly with a many different choices.)

As you can see, this can be a huge problem if you are trying to store the new augmented datasets since this increases the memory footprint by a number of multiple times. 

Please let me know if you can replicate this:
any image--> augment it with gauss noise + others and then save it. See if the footprint has increased. 

**If this is intended, why does this happen?**

**Comment by jspaezp** [comment]
what format are you using to save the images?

**Comment by marvision-ai** [comment]
@jspaezp  jpg and png

**Comment by jspaezp** [comment]
Well I think this has to do with the way the images are actually encoded by the format (aka: not in the scope of imgaug).

(this is a pretty gross over-simplification of what actually happens) Basically, uniform regions of the image can be represented more efficiently (think: from x to y its all white), whilst noisy or non uniform regions are not.

This can be exemplified like this, if we create a fake uniform image and a fake noisy one and we save both to disk

```python
import numpy as np
import cv2

blank_img = np.zeros(shape = (512,512,3), dtype = np.uint8)
noise_img = np.random.randint(255, size=(300, 600, 3), dtype = np.uint8)

cv2.imwrite("img1.png", blank_img)
cv2.imwrite("img1.jpg", blank_img)
cv2.imwrite("noise1.jpg", noise_img)
cv2.imwrite("noise1.png", noise_img)
```

```shell
! ls -lcth *.{png,jpg}
# -rw-r--r-- 1 root root 770K Sep 18 17:49 noise1.png
# -rw-r--r-- 1 root root 301K Sep 18 17:49 noise1.jpg
# -rw-r--r-- 1 root root 4.7K Sep 18 17:49 img1.jpg
# -rw-r--r-- 1 root root 1.7K Sep 18 17:49 img1.png
```

Note how the uniform images are represented in disk as < 5kb each. And even if they have the same size (512x512), the noisy image uses ~200x more space in disk (770 kb for png and 301 kb for jpg).

Therefore it is not surprising that adding noise makes images harder to store and therefore have larger file sizes

```python
import imgaug as ia
import imgaug.augmenters as iaa

aug = iaa.AdditiveGaussianNoise(scale=(0, 0.2*255))
noisy_blank = aug(images = blank_img)
noisy_noise = aug(images = noise_img)

cv2.imwrite("n_img1.png", noisy_blank)
cv2.imwrite("n_img1.jpg", noisy_blank)
cv2.imwrite("n_noise1.jpg", noisy_noise)
cv2.imwrite("n_noise1.png", noisy_noise)
```

```shell
! ls -lcth *.{png,jpg}
# -rw-r--r-- 1 root root 770K Sep 18 17:50 n_noise1.png
# -rw-r--r-- 1 root root 305K Sep 18 17:50 n_noise1.jpg
# -rw-r--r-- 1 root root 171K Sep 18 17:50 n_img1.jpg
# -rw-r--r-- 1 root root 549K Sep 18 17:50 n_img1.png
```

Note that the size of the noisy image does not change dramatically (since it is already as noisy as an image can exist) but the uniform one dramatically increases in size.

Hope this explains your issue and helps you find a workaround for your disk/memory restrictions.
Best

**Comment by marvision-ai** [comment]
@jspaezp this makes complete sense. Thanks for the awesome explanation! 

It's somewhat along the lines of what I came up with....Until I tried saving with other libraries. If you save with imageio it results in significantly smaller image footprint. Do you know why this could be?  I guess this is out of the scope for imgaug but I'm just curious if you might know why. 

Thanks  again!

**Comment by jspaezp** [comment]
@marvision-ai 

**TLDR: I think they default to different compression strategies, CV2 can get the same efficiency**

I was intrigued by that observation ... so I went into a rabbit-hole to check it out.

Looking at the implementation... It seems like `imageio` dispatches the writing to FreeImageFormat (not going to go any deeper into it ...) https://github.com/imageio/imageio/blob/22f5d58855e32cbb4789da67b9a73650983b59f3/imageio/plugins/freeimage.py#L240

Looking at the efficiency ... (if we continue with our former example...) and I save the files with imageio.
It seems like it does make smaller images (for noisy uniform png images, 396K vs 550K).

```python
import imageio

imageio.imwrite("iio_n_img1.png", noisy_blank)
imageio.imwrite("iio_n_img1.jpg", noisy_blank)
imageio.imwrite("iio_n_noise1.jpg", noisy_noise)
imageio.imwrite("iio_n_noise1.png", noisy_noise)
```

```shell
ls -lcth

# -rw-r--r-- 1 root root  61K Sep 23 04:18 iio_n_img1.jpg
# -rw-r--r-- 1 root root 396K Sep 23 04:18 iio_n_img1.png
# -rw-r--r-- 1 root root 157K Sep 23 04:18 iio_n_noise1.jpg
# -rw-r--r-- 1 root root 759K Sep 23 04:18 iio_n_noise1.png
# -rw-r--r-- 1 root root 4.7K Sep 23 04:18 img1.jpg
# -rw-r--r-- 1 root root 1.7K Sep 23 04:18 img1.png
# -rw-r--r-- 1 root root 171K Sep 23 04:18 n_img1.jpg
# -rw-r--r-- 1 root root 550K Sep 23 04:18 n_img1.png
# -rw-r--r-- 1 root root 304K Sep 23 04:18 n_noise1.jpg
# -rw-r--r-- 1 root root 770K Sep 23 04:18 n_noise1.png
# -rw-r--r-- 1 root root 301K Sep 23 04:18 noise1.jpg
# -rw-r--r-- 1 root root 770K Sep 23 04:18 noise1.png
```

NONTHELESS...
I also noticed CV2 has the option to set compression parameters to the PNG writter, so ... 

```python
for comp_level in [0, 5, 9, 16]:
    # Max compression is 9, any more will just use 9
    cv2.imwrite('compress_n_img_' + str(comp_level) + '.png', noisy_blank,  [cv2.IMWRITE_PNG_COMPRESSION, comp_level])
    cv2.imwrite('compress_n_noise_' + str(comp_level) + '.png', noisy_noise,  [cv2.IMWRITE_PNG_COMPRESSION, comp_level])
```

```shell
ls -lcth comp*.{png}
# -rw-r--r-- 1 root root 770K Sep 23 04:19 compress_n_img_0.png
...
# -rw-r--r-- 1 root root 424K Sep 23 04:19 compress_n_img_9.png
# -rw-r--r-- 1 root root 770K Sep 23 04:19 compress_n_noise_0.png
...
# -rw-r--r-- 1 root root 762K Sep 23 04:19 compress_n_noise_9.png
```

We get that on the highest compression level.
- The uniform image with noise is 424K with cv2 and 396K with imageio (IO is better).
- And for the noisy noise image, 762K (CV2) > 759K (IO is better as well)

Some plotting to visualize the differences

Comparing the size of all the images...
When we focus on only PNG's with the added noise ...
![image](https://user-images.githubusercontent.com/8116459/93969523-11653f80-fd3a-11ea-8498-62990baf54fd.png)

*Moreover* .... CV2 allows the selection of different compression strategies
Documented here: https://docs.opencv.org/4.4.0/d4/da8/group__imgcodecs.html#ga292d81be8d76901bff7988d18d2b42ac

```
 for comp_strat in range(4):
    cv2.imwrite(
        'compress_n_img_strat_'+ str(comp_strat) + '.png',
        noisy_blank,
        [cv2.IMWRITE_PNG_COMPRESSION, 9, cv2.IMWRITE_PNG_STRATEGY, comp_strat])
    cv2.imwrite('compress_n_noise_strat_'+ str(comp_strat) + '.png'',
        noisy_noise,
        [cv2.IMWRITE_PNG_COMPRESSION, 9, cv2.IMWRITE_PNG_STRATEGY, comp_strat])

```

```
# -rw-r--r-- 1 root root 367K Sep 23 05:51 compress_n_img_strat_2.png
# -rw-r--r-- 1 root root 384K Sep 23 05:51 compress_n_img_strat_3.png
# -rw-r--r-- 1 root root 396K Sep 23 05:51 compress_n_img_strat_1.png
# -rw-r--r-- 1 root root 423K Sep 23 05:51 compress_n_img_strat_0.png
# -rw-r--r-- 1 root root 505K Sep 23 06:03 compress_n_img_strat_4.png

# -rw-r--r-- 1 root root 760K Sep 23 05:51 compress_n_noise_strat_3.png
# -rw-r--r-- 1 root root 760K Sep 23 05:51 compress_n_noise_strat_2.png
# -rw-r--r-- 1 root root 760K Sep 23 05:51 compress_n_noise_strat_1.png
# -rw-r--r-- 1 root root 761K Sep 23 05:51 compress_n_noise_strat_0.png
# -rw-r--r-- 1 root root 811K Sep 23 06:03 compress_n_noise_strat_4.png
```

We get that on the highest compression, and the best strategy (for each ...).
- The uniform image with noise is 384K with cv2 and 396K with imageio (CV2 is better).
- And for the noisy noise image, 760K (CV2) > 759K (IO marginally better).

Therefore, it is likely that `imageio` just defaults to a more optimal strategy for your
kind of images (not sure) and a higher compression level (of 9) than the default of 1 in CV2.

Note that none of these are free changes, some have implication in read/write speed
at the expense of disk usage... just find what trade-off is best for your application.

Best!

**Comment by marvision-ai** [comment]
@jspaezp  wow that's amazing. What an investigation! Thanks for doing that. 

PS. You should remove the TLDR. This stuff is fascinating. haha
